runtime/src/main/java/org/apache/flink/agents/runtime/feedback/Checkpoints.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.agents.runtime.feedback;
+
+import org.apache.flink.agents.runtime.logger.FeedbackLogger;
+import org.apache.flink.util.IOUtils;
+
+import java.io.OutputStream;
+import java.util.Objects;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.Supplier;
+
+/**
+ * NOTE: This source code was copied from the <a
+ * href="https://github.com/apache/flink-statefun">flink-statefun</a>
```

Comment:
Looks we copied lots of code from other projects directly, what's the main reason for copying/what did we modify after copying? Will there be problem if we need to import their lib later?